### PR TITLE
[None][feat] Add Gemma4 variable-window attention

### PR DESCRIPTION
**Pull request by achartier**
## Dev Engineer Review

- Adds variable-window attention for Gemma4 multimodal context requests on SM100/SM103 GPUs.
- Threads inclusive per-query KV bounds through Python, nanobind, C++, FMHA dispatch, TRTLLM-Gen metadata, and kernel parameters.
- Preserves causal behavior for full-attention layers and expands image/video ranges for sliding-attention layers.
- Replaces one attention-window size with independent left and right reaches.
- Routes variable-window context requests to TRTLLM-Gen and retains FlashInfer for eligible generation requests.
- Validates paired, contiguous, `int32`, context-only bound tensors.
- Extends mask classification, support probing, XQA routing, and FMHA cache keys.
- Updates TRTLLM-Gen libraries and Blackwell cubins. Mixed FP8/NVFP4 handling, CUDA launch attributes, and kernel resource changes increase ABI and performance risk.
- Main risks are interface alignment, packed multimodal bound calculation, binary compatibility, XQA routing, and standard-attention regressions.

## QA Engineer Review

- Adds unit coverage for FlashInfer fallback, FMHA cache separation, PrimTS rejection, Gemma4 routing, variable-window propagation, multimodal input handling, and mask serialization.
- No integration test-list files changed.
- Coverage is **needs follow-up** because direct TRTLLM-Gen execution, FMHA support probing, XQA routing, and all tensor-validation paths are not fully covered by the changed tests.

## Per-File QA Perspective

- `cpp/tensorrt_llm/common/attentionOp.cpp`, `attentionOp.h`: Add variable-window state and runtime parameters. Verify context and generation routing.
- `cpp/tensorrt_llm/kernels/contextFusedMultiHeadAttention/fused_multihead_attention_common.h`: Add the variable-window mask and bounds. Verify defaults and serialization.
- `cpp/tensorrt_llm/kernels/fmhaDispatcher.cpp`: Map masks, probe support, validate window combinations, and forward bounds. Verify safe rejection.
- `cpp/tensorrt_llm/kernels/trtllmGenKernels/fmha/*`: Add variable-window metadata, validation, mask classification, bound propagation, mixed-dtype handling, and launch configuration. Verify ABI, pointer defaults, dtype paths, CUDA-version paths, and kernel results.
- `cpp/tensorrt_llm/kernels/trtllmGenKernels/fmha/lib/x86_64-linux-gnu/*.a`: Update TRTLLM-Gen binaries. Verify symbol, ABI, and GPU compatibility.
- `cpp/tensorrt_llm/kernels/xqaDispatcher.cpp`: Change XQA mask and window routing. Verify chunked, sliding, causal, and speculative paths.
- `cpp/tensorrt_llm/nanobind/thop/bindings.cpp`, `cpp/tensorrt_llm/thop/attentionOp.*`: Expose and validate bound tensors. Verify argument order, device, phase, dtype, shape, and contiguity checks.
- `tensorrt_llm/_torch/attention/backends/*`: Add bound tensors, backend rejection, fallback forwarding, architecture gating, and cache isolation. Verify incomplete-bound rejection and selection behavior.
- `tensorrt_llm/_torch/models/modeling_gemma4.py`: Compute and propagate packed image/video ranges. Verify batch offsets, layer routing, mask equivalence, and full-attention preservation.
- `tensorrt_llm/_torch/models/modeling_gemma4mm.py`, `modeling_multimodal_mixin.py`: Pass attention metadata through multimodal forwards. Verify caller compatibility.
- `tests/unittest/_torch/attention/test_combined_fmha.py`: Covers FlashInfer context rejection and generation support. No integration test-list entry changed.
- `tests/unittest/_torch/attention/test_fmha_manager.py`: Covers cache separation. No integration test-list entry changed.
- `tests/unittest/_torch/attention/test_prims_ts_fmha.py`: Covers PrimTS rejection. No integration test-list entry changed.
- `tests/unittest/_torch/attention/test_variable_window_validation.py`: Covers device mismatch and unsupported FMHA configuration errors on supported GPU architectures. No integration test-list entry changed.
- `tests/unittest/_torch/modeling/test_modeling_gemma4.py`: Covers Gemma4 propagation, routing, mask equivalence, SM107 behavior, and generation behavior. No integration test-list entry changed.
- `tests/unittest/_torch/modeling/test_gemma4_multimodal.py`: Covers multimodal input and attention-metadata forwarding. No integration test-list entry changed.
- `cpp/tests/unit_tests/kernels/trtllmGenExportCompatTest.cpp`: Verifies `VariableWindow` serialization. No integration test-list entry changed.
- `cpp/tests/unit_tests/kernels/CMakeLists.txt`: Adds the TRTLLM-Gen compatibility test target. Verify that the target builds and runs in the unit-test workflow.

## Description

Add end-to-end TRTLLM-Gen variable-window context attention for Gemma4 multimodal bidirectional masks on Blackwell.

The change:

- builds inclusive per-query KV start/end bounds for contiguous image and video token blocks;
- applies those bounds only to Gemma4 sliding-window layers while full-attention layers remain causal;
- threads the bounds through the FMHA backend, THOP binding, attention runtime, dispatcher, and TRTLLM-Gen interface;
- translates existing causal and bidirectional sliding-window semantics to explicit left/right reaches;
- falls back from the pinned FlashInfer context implementation to native TRTLLM-Gen when variable-window bounds are present, while retaining FlashInfer for eligible generation phases;
- validates the context-only bound tensors and keeps FMHA implementation selection cache-safe; and
- imports the TRTLLM-Gen interface/library update plus 108 SM100-family variable-window cubins.

The implementation was informed by the earlier `nvrohanv/add_bidirectional_and_gemma_trtllm_gen` integration branch, adapted to the current attention backend and generated ABI.

## Test Coverage

- Clean native Release build on an NVIDIA B200:
  `python3 scripts/build_wheel.py --install -b Release -a native --clean --nvtx --use_ccache --build_dir cpp/build`
- Gemma4 variable-window mask equivalence, Blackwell routing, and sliding-layer propagation: 3 passed.
- FlashInfer/TRTLLM-Gen context fallback matrix: 144 passed.
- Attention Python/C++ boundary synchronization plus combined FMHA tests: 163 passed.
- Direct TRTLLM-Gen VariableWindow FMHA run on B200 with full reference checking: passed.
- Repository pre-commit hooks: passed.

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- If PR introduces API changes, an appropriate PR label is added - either `api-compatible` or `api-breaking`. For `api-breaking`, include `BREAKING` in the PR title.
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

To see a list of available CI bot commands, please comment `/bot help`.

### CODE REVIEW

File: cpp/tensorrt_llm/kernels/fmhaDispatcher.cpp [hunk 1]
```diff
@@ -242,8 +275,53 @@ void FmhaDispatcher::run(MHARunnerParams runnerParams)
...
+            tllmRunnerParams.mMaskType = TrtllmGenAttentionMaskType::VariableWindow;
+        }
+        else if (usesChunkedAttention)
+        {
+            tllmRunnerParams.mMaskType = TrtllmGenAttentionMaskType::SlidingOrChunkedCausal;
+            tllmRunnerParams.mChunkedAttentionSize = runnerParams.chunkedAttentionSize;
+        }
+        else if (mFixedParams.attentionMaskType == ContextAttentionMaskType::BIDIRECTIONAL_SLIDING_WINDOW)
+        {
+            auto const windowReach = runnerParams.slidingWindowSize / 2;
```
[the remark this answers is not in the document]
Comment (achartier) [reply cpp/tensorrt_llm/kernels/fmhaDispatcher.cpp:302]: addressed by 48c01bd75135d944a2f4d62621c2ffc7da78b51d

File: cpp/tensorrt_llm/kernels/trtllmGenKernels/fmha/trtllmGen_fmha_export/KernelConfigBase.h [hunk 2]
```diff
@@ -401,6 +428,8 @@ template <> inline std::string toString(AttentionMaskType e) {
     return "Custom";
   case AttentionMaskType::SlidingWindowCustom:
     return "SlidingWindowCustom";
+  case AttentionMaskType::VariableWindow:
+    return "VariableWindow";
```
[the remark this answers is not in the document]
Comment (achartier) [reply cpp/tensorrt_llm/kernels/trtllmGenKernels/fmha/trtllmGen_fmha_export/KernelConfigBase.h:432]: addressed by 48c01bd75135d944a2f4d62621c2ffc7da78b51d

File: cpp/tensorrt_llm/thop/attentionOp.cpp [hunk 3]
```diff
@@ -1440,6 +1452,25 @@ void attention(torch::Tensor q, std::optional<torch::Tensor> k, std::optional<to
...
+    };
+    if (variable_window_token_starts.has_value())
+    {
+        auto const smVersion = tensorrt_llm::common::getSMVersion();
+        TORCH_CHECK(smVersion == 100 || smVersion == 103,
+            "variable-window attention requires an SM100 or SM103 GPU, got SM", smVersion);
+        TORCH_CHECK(num_contexts > 0, "variable-window attention requires a context phase");
+        validateVariableWindowTensor(variable_window_token_starts.value(), "variable_window_token_starts");
+        validateVariableWindowTensor(variable_window_token_ends.value(), "variable_window_token_ends");
+    }
```
[the remark this answers is not in the document]
Comment (achartier) [reply cpp/tensorrt_llm/thop/attentionOp.cpp:1478]: addressed by 48c01bd75135d944a2f4d62621c2ffc7da78b51d

File: tensorrt_llm/_torch/attention/backends/fmha/manager.py [hunk 4]
```diff
@@ -133,6 +133,7 @@ class _FmhaCacheKey(NamedTuple):
     generation_batch_size: int
     generation_seq_len_q: int
     attention_mask_type: AttentionMaskType
+    has_variable_window: bool
```
[the remark this answers is not in the document]
Comment (achartier) [reply tensorrt_llm/_torch/attention/backends/fmha/manager.py:136]: addressed by 48c01bd75135d944a2f4d62621c2ffc7da78b51d

File: tensorrt_llm/_torch/attention/backends/fmha/prims_ts.py [hunk 5]
```diff
@@ -254,6 +254,11 @@ def _is_supported_with_reason(
             return False, "attention sinks are not supported."
         if fwd.attention_mask_data is not None:
             return False, "custom attention masks are not supported."
+        if (
+            fwd.variable_window_token_starts is not None
+            or fwd.variable_window_token_ends is not None
+        ):
+            return False, "variable-window attention is not supported by this adapter."
```
[the remark this answers is not in the document]
Comment (achartier) [reply tensorrt_llm/_torch/attention/backends/fmha/prims_ts.py:261]: addressed by 48c01bd75135d944a2f4d62621c2ffc7da78b51d

File: tensorrt_llm/_torch/models/modeling_gemma4.py [hunk 6]
```diff
@@ -1499,13 +1598,29 @@ def forward(
...
         # applies multimodal bidirectionality only to sliding-window layers;
         # full-attention layers retain their normal causal mask.
         use_bidir = getattr(self.config, "use_bidirectional_attention", None)
-        if mm_token_type_ids is not None and use_bidir == "vision":
+        if (
+            mm_token_type_ids is not None
+            and use_bidir == "vision"
+            and is_sm_100f()
+            and isinstance(attn_metadata, TrtllmAttentionMetadata)
+        ):
```
[the remark this answers is not in the document]
Comment (achartier) [reply tensorrt_llm/_torch/models/modeling_gemma4.py:1626]: unsupported architecture/metadata already uses the dense-mask fallback, while request-specific support is correctly evaluated later by FmhaManager.select() with complete inputs

[the remark this answers is not in the document]
Comment (achartier) [reply tensorrt_llm/_torch/models/modeling_gemma4.py:1626]: fixed in 753024cd80

File: cpp/tensorrt_llm/kernels/trtllmGenKernels/fmha/trtllmGen_fmha_export/trtllmGenExportCompat.h [hunk 7]
```diff
@@ -156,6 +162,8 @@ template <> inline std::string toString(AttentionMaskType e) {
     return "Custom";
   case AttentionMaskType::SlidingWindowCustom:
     return "SlidingWindowCustom";
+  case AttentionMaskType::VariableWindow:
+    return "VariableWindow";
```
[the remark this answers is not in the document]
Comment (achartier) [reply cpp/tensorrt_llm/kernels/trtllmGenKernels/fmha/trtllmGen_fmha_export/trtllmGenExportCompat.h:166]: fixed in 753024cd80

File: tensorrt_llm/_torch/models/modeling_gemma4mm.py [hunk 8]
```diff
@@ -944,7 +945,11 @@ def get_language_model_extra_forward_kwargs(
         """Build Gemma4-specific language-model forward arguments."""
         del position_ids, forward_kwargs
         mm_token_type_ids = None
-        if raw_input_ids is not None and mm_inputs.input_ids is None:
+        if (
+            attn_metadata.num_contexts > 0
+            and raw_input_ids is not None
+            and mm_inputs.input_ids is None
+        ):
```
[the remark this answers is not in the document]
Comment (achartier) [reply tensorrt_llm/_torch/models/modeling_gemma4mm.py:952]: fixed in 753024cd80

File: tensorrt_llm/_torch/models/modeling_multimodal_mixin.py [hunk 9]
```diff
@@ -726,6 +726,7 @@ def get_language_model_forward_kwargs(
         }
         llm_kwargs.update(
             self.get_language_model_extra_forward_kwargs(
+                attn_metadata=attn_metadata,
```
[the remark this answers is not in the document]
Comment (achartier) [reply tensorrt_llm/_torch/models/modeling_multimodal_mixin.py:729]: fixed in 753024cd80

**Comment by yuxianq** [comment]
@achartier please split cubins update into a separate commit, so that we can review the rest files easily. Also don't forget to update https://gitlab-master.nvidia.com/jdemouth/trtllm-gen/-/wikis/TRTLLM-Gen-FMHA-doc#track-of-the-commits if this PR merged. Thanks~
